gcovr/writer/html.py

```diff
@@ -367,34 +384,47 @@ def print_html_report(covdata: CovData, output_file, options):
         data["css_link"] = css_link
 
     data["theme"] = options.html_theme
+    data["row_sort_key"], data["row_sort_reverse"] = row_sort_keys(options)
 
     root_info.set_coverage(covdata)
 
     # Generate the coverage output (on a per-package basis)
     # source_dirs = set()
     files = []
-    dirs = []
     filtered_fname = ""
-    keys = sort_coverage(
+    sorted_keys = sort_coverage(
         covdata,
         show_branch=False,
         filename_uses_relative_pathname=True,
         by_num_uncovered=options.sort_uncovered,
         by_percent_uncovered=options.sort_percent,
     )
+
+    if options.html_cascaded_directories:
+        root_info.subdirs = DirectoryCoverage.from_covdata(
+            covdata, sorted_keys, options.root_filter
+        )
+        DirectoryCoverage.collapse_subdirectories(
+            root_info.subdirs, options.root_filter
+        )
+
     cdata_fname = {}
     cdata_sourcefile = {}
-    for f in keys:
+    for f in list(covdata.keys()) + list(root_info.subdirs.keys()):
         filtered_fname = options.root_filter.sub("", f)
         files.append(filtered_fname)
-        dirs.append(os.path.dirname(filtered_fname) + os.sep)
         cdata_fname[f] = filtered_fname
         if options.html_details:
-            cdata_sourcefile[f] = _make_short_sourcename(output_file, filtered_fname)
+            if os.path.normpath(f) == os.path.normpath(options.root_dir):
+                cdata_sourcefile[f] = output_file
+            else:
+                cdata_sourcefile[f] = _make_short_sourcename(
+                    output_file, filtered_fname
+                )
```

Comment:
Why was this changed?

Reply:
The "output_file" parameter needs to be preserved for the top-most directory report we need to give. Otherwise, we need to manage a regular _make_short_sourcename.

Reply:
In the old version `f` was always a source file. Can it now also be the root directory?

Reply:
Yes. Or any subdirectory. They are each entries with their own statistics.
(exception, if there are directories that have only 1 entry they are collapsed for easier navigation)

Reply:
Ok. In this case also `filtered_fname` should be an empty string.
If this isn't true because of the trailing slash in the `options.root_filter` this should be handled in line 414.
Here we can then rely on the value of `filtered_fname`.